### PR TITLE
Add dropwizard metrics to SnapshotTransaction

### PR DESCRIPTION
**Pull request by lmaPalantir**
---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1356)

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/atlasdb](https://github.com/palantir/atlasdb), @lmaPalantir! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: atlasdb-commons/src/main/java/com/palantir/util/AtlasDbMetrics.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    private static final String DEFAULT_REGISTRY_NAME = "AtlasDb";
+    private static final int DEFAULT_LOG_PERIOD_SECONDS = 60;
+    private static volatile AtlasDbMetrics instance = null;
+
+    private MetricRegistry metricRegistry;
+    private final Logger metricsLogger;
+    private final Slf4jReporter metricsLogReporter;
+
+    private AtlasDbMetrics(String logDomain, String registryName) {
+        this.metricRegistry = SharedMetricRegistries.getOrCreate(registryName);
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/util/AtlasDbMetrics.java]: I want to be able to inject the MetricRegistry used, for example one from DropWizard's Environment#getMetrics()

File: atlasdb-commons/src/main/java/com/palantir/util/AtlasDbMetrics.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        JmxReporter.forRegistry(metricRegistry).inDomain(logDomain).convertDurationsTo(
+                TimeUnit.MILLISECONDS).convertRatesTo(TimeUnit.SECONDS).build().start();
+        this.metricsLogger = LoggerFactory.getLogger(logDomain);
+        this.metricsLogReporter = Slf4jReporter.forRegistry(metricRegistry).outputTo(metricsLogger).convertRatesTo(
+                TimeUnit.SECONDS).convertDurationsTo(TimeUnit.MILLISECONDS).build();
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/util/AtlasDbMetrics.java]: We should probably separate initialization of reporters as some registries will already be reporting (e.g. DropWizard automatically enables JMX + any configured in service configuration).

File: atlasdb-commons/build.gradle [hunk 3]
```diff
@@ -19,6 +19,7 @@ dependencies {
     compile project(":commons-executors")
     compile 'javax.ws.rs:javax.ws.rs-api:2.0.1'
     compile group: 'com.fasterxml.jackson.core', name: 'jackson-annotations'
+    compile group: 'com.codahale.metrics', name: 'metrics-core'
```
Comment (schlosna) [inline atlasdb-commons/build.gradle]: This is going to cause conflicts with folks that depend on newer versions of DW metrics via 'io.dropwizard.metrics:metrics-core'

File: atlasdb-commons/src/main/java/com/palantir/util/AtlasDbMetrics.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.codahale.metrics.JmxReporter;
+import com.codahale.metrics.MetricRegistry;
+import com.codahale.metrics.SharedMetricRegistries;
+import com.codahale.metrics.Slf4jReporter;
+
+public class AtlasDbMetrics {
```
Comment (schlosna) [inline atlasdb-commons/src/main/java/com/palantir/util/AtlasDbMetrics.java]: I think we'll actually want this in the `atlasdb-client` module, and we should simplify this as just a holder for the `MetricRegistry` and we should rely on the user handling all of the reporting and logging of metrics outside of AtlasDB. Basically I envisage just adding the following first, then we can wrap things like Transaction

```
package com.palantir.atlasdb.metrics;

import com.codahale.metrics.MetricRegistry;
import com.google.common.base.Preconditions;

public final class AtlasDbMetrics {

    private static volatile MetricRegistry metrics;

    private AtlasDbMetrics() { }

    public static void setMetrics(MetricRegistry metricRegistry) {
        Preconditions.checkNotNull(metricRegistry, "metric registry cannot be null");
        AtlasDbMetrics.metrics = metricRegistry;
    }

    public static MetricRegistry metrics() {
        MetricRegistry metricRegistry = AtlasDbMetrics.metrics; // single volatile read
        Preconditions.checkState(metricRegistry != null, "metric registry has not been initialized");
        return metricRegistry;
    }

}
```

**Comment by lmaPalantir** [comment]
new pull request here https://github.com/palantir/atlasdb/pull/1429

**Comment by gsheasby** [comment]
Superceded by #1429

**Pull request closed without merging**